### ISSUE TITLE
prow: clonerefs: speed up cloning of large repos through shallow cloning / pre-cloned repo in base image

### ISSUE THREAD

**Issue by AviKndr** [kubernetes/test-infra#10388]
**What would you like to be added**:
I would like to add some feature to alleviate the overhead of cloning on every job for very large repos (> 1 GB). 

Given that we only need the specific revision and shallow cloning for our repo takes about half of the time, we'd like to add a "--shallow" arg to clonerefs for cloning without tags/history (depth of 1) as referenced here: #6865 . 

Alternatively, a way of preserving the source by having a checkout image with a pre-cloned repo used as the base on which to fetch new changes for jobs.

**Why is this needed**:
Currently our presubmit jobs take ~13 minutes in which ~10 are a result of cloning the entire repo (Linux Kernel). This would be very inefficient in terms of both speed and network usage with a large number of concurrent presubmit jobs for any kernel related jobs.

/kind feature
/area prow

**Comment by AviKndr** [comment]
@BenTheElder

**Comment by stevekuznetsov** [comment]
There are a large number of tools inside of repos that expect to have deep clones -- the k8s versioning bash stuff is a prime example that needs to know about tags, commits between them, etc. A shallow clone option might be nice but would start to leak infra state into the job which isn't ideal.

On our end we do use a cache container image to make cloning faster, but that's entirely external to the utilities and just something to set up in the job configuration.

@BenTheElder had some ideas about cloning into a volume that held a cache, but that imposes limits on job concurrency while maintaining the caching. Not sure what the final idea was there

**Comment by BenTheElder** [comment]
I suggested shallow cloning would be a reasonable and simple option, it should not be the default but we could allow enabling shallow clones for projects that don't need full clones.

AFAICT, they could get away with shallow clones, other projects probably could too, and flipping a switch to enable that will be simpler than the pre-cloned repo support.

**Comment by stevekuznetsov** [comment]
/remove-lifecycle stale

An option to do a shallow clone would be great

/help

**Comment by vishleshpatel** [comment]
@stevekuznetsov if its okay for a newbie to learn system and work on it, then I would like to go ahead and pick this one up.

**Comment by stevekuznetsov** [comment]
@vishleshpatel hi! Totally :)

Job configuration holds information about the references that are under test and need to be cloned as well as any extra references that might be necessary for the job. Options for the the refs under test are in the `UtilityConfig`:

https://github.com/kubernetes/test-infra/blob/143a95a43bb70d540e021e23fd50db5020b30089/prow/config/jobs.go#L395-L406

You could add something like:

```go
// CloneDepth is the depth of the clone that will be used. A depth of zero will do a full clone.
CloneDepth int `json:"clone_depth,omitempty"`
```

However, to add the same option for the refs that are not under test, you'd need to add it to `Refs` [here](https://github.com/kubernetes/test-infra/blob/143a95a43bb70d540e021e23fd50db5020b30089/prow/apis/prowjobs/v1/types.go#L459-L486). The program we use to clone is [`clonerefs`](https://github.com/kubernetes/test-infra/tree/master/prow/cmd/clonerefs) and it will read the `Refs`, where you can decide how deep to clone.

Check out [this commit](https://github.com/kubernetes/test-infra/commit/535a73e2beb4645c6209eab624036eb03655bf4a) for an example of adding a change like this!

**Comment by vishleshpatel** [comment]
Thanks @stevekuznetsov for the pointers. I will take a look at it this week and ask questions if in need.

**Comment by vishleshpatel** [comment]
Hi @stevekuznetsov 

so i now have changes put together. How to test the changes? Also, What tests needs to be run?

Thanks!

**Comment by stevekuznetsov** [comment]
This got implemented in #12499

/close
